audit/base/base.go

```diff
@@ -0,0 +1,556 @@
+	default:
+		return Error{Err: errors.New("queue is full"), Category: ErrQueueFull}
+	}
+
+	// If we had any errors previously, we need to return them.
+	if err := c.getErr(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Reset resets the connection to the audit server.
+// This can be used if the connection is in a bad state and needs to be reset.
+// This will cause an existing connection to be closed and reset internal state.
+// This is thread safe.
+func (c *Client) Reset(ctx context.Context, newConn conn.Audit) error {
+	if c == nil {
+		return Error{Err: fmt.Errorf("cannot call Reset on a base.Client that is nil"), Category: ErrConnection}
+	}
+	if newConn == nil {
+		return Error{Err: fmt.Errorf("cannot pass a conn.Audit that is nil"), Category: ErrConnection}
+	}
+
+	// If for some reason there is no error happening, we need to set one to prevents sends.
+	c.setErr(Error{Err: fmt.Errorf("audit client Reset() called, resetting connection"), Category: ErrConnection})
+
+	// If the connection is live, we need to close it.
+	c.close(ctx, false)
+	c.conn.Store(&newConn)
+
+	// Now let's bring everything back to life.
+	c.done = make(chan struct{})
+
+	c.closeOnceMu.Lock()
+	c.closeOnce = sync.Once{}
+	c.closeOnceMu.Unlock()
+
+	c.setErr(nil)
+	return nil
+}
+
+// Close closes the connection to the audit server.
+func (c *Client) Close(ctx context.Context) error {
+	c.wait()
+	return c.close(ctx, true)
+}
+
+// wait waits for the client to finish sending all messages. This is determined by the queue
+// being empty.
+func (c *Client) wait() {
+	for {
+		if len(c.sendCh) == 0 {
+			return
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+}
+
+// close handles closing the connection to the audit server. This is used by Close().
+// It ensures that the connection is only closed once.
+func (c *Client) close(ctx context.Context, stopSender bool) error {
+	c.closeOnceMu.Lock()
+	defer c.closeOnceMu.Unlock()
+
+	c.closeOnce.Do(func() {
+		if stopSender {
+			// This might be new to some of you, but you can send a channel over a channel.
+			// In this case, I create a channel and send it on the stopSend channel. When the
+			// sender receives the channel, it will close the channel I sent, and then we know that
+			// the sender is done.
+			sig := make(chan struct{}, 1)
+			c.stopSend <- sig
+			<-sig
+		}
+
+		if err := c.getErr(); err != nil {
+			ptr := c.conn.Load()
+			if ptr != nil {
+				(*ptr).CloseSend(ctx) // Ignore any error
+			}
+			return
+		}
+
+		ptr := c.conn.Load()
+		if ptr != nil {
+			if err := (*ptr).CloseSend(ctx); err != nil {
+				c.setErr(Error{Err: err, Category: ErrConnection})
+			}
+		}
+	})
+	return c.getErr()
+}
+
+// sender is the async sender for the audit client.
+func (c *Client) sender() {
+	defer close(c.done)
+
+	var ticker *time.Ticker
+	defer func() {
+		if ticker != nil {
+			ticker.Stop()
+		}
+	}()
+
+	var tickerCh <-chan time.Time
+
+	for {
+		if c.successSend && ticker == nil {
```

Comment:
I think this is only sending a heartbeat once we've successfully sent a message to the server, but this supposes that the user has called `Send` at least once before heartbeats start (i.e. they don't start when the client is created), is that right?

If that's right, I'm not sure that actually meshes with my understanding of the point of a heartbeat. For example what if I'm running a web service and using this to manage my auditing... I might want to write a monitor on the heartbeat being present, but if I stand my webserver up in 60 regions and 10 of those regions have low API usage (== no audits) then my heartbeats wont start and I might end up getting alerted that something is broken.

It _feels_ like, if the client is initialized, we should start sending heartbeats immediately, not wait for the first send?

Unless I'm misunderstanding what this code is doing...

Reply:
This is a design decision that is out of my hands.  This is how the C# and C++ clients operate.  I don't agree with the decision, however the reasoning is:

Services will try and not implement this and just use the heartbeat to say they are compliant. Therefore, until they successfully send a message, they don't get a heartbeat.

This is problematic reasoning in several different ways, including what you are talking about.  And it is probably worth writing a document on improvements to existing or v2 versions of this.  But unfortunately this is how v1 works.